src/runtime-rs/crates/runtimes/virt_container/src/container_manager/process.rs

```diff
@@ -78,7 +81,44 @@ impl Process {
             exit_watcher_rx: Some(receiver),
             exit_watcher_tx: Some(sender),
             wg_stdin: WaitGroup::new(),
+            passfd_io: None,
+        }
+    }
+
+    pub async fn passfd_io_init(&mut self, hvsock_uds_path: &str, passfd_port: u32) -> Result<()> {
+        let mut passfd_io =
+            PassfdIo::new(self.stdin.clone(), self.stdout.clone(), self.stderr.clone()).await;
+
+        passfd_io
+            .open_and_passfd(hvsock_uds_path, passfd_port, self.terminal)
+            .await
+            .context("passfd connect")?;
+
+        self.passfd_io = Some(passfd_io);
+
+        Ok(())
+    }
+
+    pub async fn passfd_io_start(&mut self, agent: Arc<dyn Agent>) -> Result<()> {
+        let wg = WaitGroup::new();
+        let wgw = wg.worker();
+
+        // The stdin, stdout, stderr are all passed to the hvsock (by passfd),
+        // so we need a normal stream as a `signal` to get notified when the
+        // guest side close io.
+        if let Some(passfd_io) = &mut self.passfd_io {
```

Comment:
Maybe we didn't need this signal stream to wait, since we have no temporary io copy in the io stream path and we just need it in the agent to drain out the pipe when no tty set.

Reply:
Not sure if I understand correctly. The signal stream is mainly used for the closing procedure rather than draining out the pipe. Without it, when the container process exited, the runtime-rs won't know that and will wait forever.

In the original io implementation(`start_io_and_wait`), "all io streams disconnect" is the condition that triggers runtime-rs to send a `WaitProcessRequest` to the agent and update the container process's state. In the passfd case, we need some equivalent trigger and I came up with the solution of the signal stream.